File: algorithms/golang/151.颠倒字符串中的单词.go
/*
 * @lc app=leetcode.cn id=151 lang=golang
 *
 * [151] 颠倒字符串中的单词
 */

// @lc code=start
func split(s string) []string {
	ret := []string{}
	for start := 0; start < len(s); {
		for start < len(s) && s[start] == ' ' {
			start++
		}
		if start >= len(s) {
			break
		}
		end := start
		for end < len(s) && s[end] != ' ' {
			end++
		}
		ret = append(ret, s[start:end])
		start = end

	}
	return ret
}

func reverseArr(arr []string) {
	for l, r := 0, len(arr)-1; l < r; {
		arr[l], arr[r] = arr[r], arr[l]
		l++
		r--
	}
}
func join(list []string, sep string) string {
	ret := ""
	for i, str := range list {
		if i > 0 {
			ret += sep
		}
		ret += str
	}
	return ret
}
func reverseWords(s string) string {
	list := split(s)
	reverseArr(list)
	return join(list, " ")
}

// @lc code=end

File: algorithms/js/151.颠倒字符串中的单词.js
/*
 * @lc app=leetcode.cn id=151 lang=javascript
 *
 * [151] 颠倒字符串中的单词
 */

// @lc code=start
/**
 * @param {string} s
 * @return {string}
 */
var reverseWords = function(s) {
  const arr = split(s);
  reverse(arr);
  return join(arr, " ");
};

function split(str) {
  const arr = [];
  const len = str.length;
  for (let start = 0; start < len; start) {
    while (start < len && str[start] == " ") {
      start++;
    }
    let end = start;
    while (end < len && str[end] != " ") {
      end++;
    }
    if (start < len) {
      arr.push(str.slice(start, end));
    }
    start = end;
  }
  return arr;
}
function join(arr, separator) {
  let ret = "";
  for (let i = 0; i < arr.length; i++) {
    if (i > 0) {
      ret += separator;
    }
    ret += arr[i];
  }
  return ret;
}
function reverse(arr) {
  for (let start = 0, end = arr.length - 1; start < end; start++, end--) {
    const tmp = arr[start];
    arr[start] = arr[end];
    arr[end] = tmp;
  }
}
// @lc code=end
